Add gRPC interop tests

JamesNK opened the pull request:
Addresses https://github.com/aspnet/AspNetCore/issues/11064

PR adds gRPC interop tests.

- Code is in src/Grpc directory
- Code is "disconnected" from the rest of build infrastructure through blank directory props/targets. If required the test project could be made to integrate with the rest of the build by moving the blank directory props/targets files to testassets.
- The interop client and server are completely independent projects, invoked by tests with dotnet run
- Tests run in a minute on my machine. Could speed them up by publishing client and running exe rather than calling dotnet run 18 times. I have kept things simple for now.
- Server runs with Kestrel. In the future we'll want to pass args to start server with HttpSys

Is this a good approach?
What changes need to happen elseware to run test project in ProjectTemplates job?

JamesNK committed:
WIP add interop tests

JamesNK committed:
Clean up

dougbu commented on src/Grpc/Directory.Build.props:
```diff
@@ -0,0 +1,2 @@
+<Project>
```
Why is it necessary to completely isolate these directories (not just test/testassets) from the repo's configuration? concern here is maintenance headache of spreading the gRPC versions around. Even for test/testassets, we usually _generate_ a file containing the versions from eng/Versions.props i.e. those projects are _mostly_ isolated from the rest of the respo.

nit: If these Directory.Build.* files are truly required, use `<Project />`.

JamesNK replied:
Update test project to not be completely isolated.

Client/server use shared runtime.

dougbu commented on src/Shared/Process/ProcessEx.cs:
```diff
@@ -0,0 +1,210 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Why is this file shared? Seems to be used in only one project.

JamesNK replied:
I took it from /src/ProjectTemplates/Helpers. Will update it to use the shared file before merging.

JamesNK replied:
Deleted file from templates project so they use the same shared file.

dougbu commented on src/Grpc/test/InteropTests/Infrastructure/ClientProcess.cs:
```diff
@@ -0,0 +1,60 @@
+using Microsoft.AspNetCore.Internal;
+using Xunit.Abstractions;
+
+namespace InteropTests.Infrastructure
```
Woah, this is the start of a lot of new test infrastructure that doesn't seem to come from the gRPC repo. Don't we have anything similar already?

JamesNK replied:
I reuse ProcessEx. ClientProcess and WebServerProcess are help types on top of that.

dougbu commented on src/Grpc/test/InteropTests/InteropTestsFixture.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace InteropTests
+{
+    public class InteropTestsFixture : IDisposable
```
Why isn't the fixture in the `InteropTests.Infrastructure` namespace?

JamesNK replied:
Moved

JamesNK committed:
Move test project into build infrastructure

JamesNK commented on src/Grpc/test/InteropTests/Helpers/InteropTestsFixture.cs:
```diff
@@ -0,0 +1,33 @@
+                return;
+            }
+
+            var webPath = @"C:\Development\Source\AspNetCore\src\Grpc\test\testassets\InteropTestsWebsite\";
```
What is the best way to find the client/server projects at runtime?

JunTaoLuo replied:
Pass it in via an assembly attribute from MSBuild: https://github.com/aspnet/AspNetCore/blob/master/src/ProjectTemplates/test/ProjectTemplates.Tests.csproj#L89-L92

JamesNK committed:
Clean up

JamesNK committed:
Remove hardcoded paths

JamesNK commented:
@Pilchie Some of the files in this PR are from grpc-dotnet. They all exist under `\src\Grpc\test\testassets`. None of this code runs outside of testing, and it doesn't ship.

What third-party-code actions are required?

JamesNK committed:
Clean up

JamesNK committed:
Fix build

JamesNK commented:
The build passes. What are the next steps here? Is it a matter of merging the PR then updating project template job tasks to add a task to run test project? Or are there changes required to the other parts of the build to handle these new projects under /src/grpc?

I don't know what changes are required from here on. I need build SMEs to guide me 🙏

Pilchie commented:
Ideally, you place the external code in a top level folder that makes it clear that it is external, and a notice added to https://github.com/aspnet/AspNetCore/blob/master/THIRD-PARTY-NOTICES.txt.

JamesNK committed:
Add copyright notices

jkotalik commented on src/Grpc/test/InteropTests/InteropTests.cs:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+
+        #region TestData
```
Ooo, spicy. I'll let @davidfowl get mad at you.

jkotalik reviewed (approved):
Don't see any reason to block this.

JamesNK committed:
Update azure template

JamesNK committed:
Fix build

JamesNK committed:
Fix build?

JamesNK committed:
Fix build?

JamesNK commented:
@JunTaoLuo Some outstanding errors. How do I resolve these.

Build error:

> C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.Managed.Core.targets(102,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true [F:\workspace.9\_work\1\s\src\Grpc\test\testassets\InteropClient\InteropClient.csproj]
> ##[error]C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.Managed.Core.targets(102,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true

Test error (one sample, this was thrown for for every test):

```
System.TimeoutException : The operation at /_/src/Grpc/test/InteropTests/InteropTests.cs:43 timed out after reaching the limit of 30000ms.
   at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter(Task task, TimeSpan timeout, String filePath, Int32 lineNumber)
   at InteropTests.InteropTests.InteropTestCase(String name) in /_/src/Grpc/test/InteropTests/InteropTests.cs:line 45
--- End of stack trace from previous location ---
Output:
[ERROR] It was not possible to find any compatible framework version
[ERROR] The framework 'Microsoft.NETCore.App', version '5.0.0-alpha1.19521.2' was not found.
[ERROR]   - The following frameworks were found:
[ERROR]       5.0.0-alpha.1.19551.1 at [F:\workspace.9\_work\1\s\.dotnet\shared\Microsoft.NETCore.App]
[ERROR]       5.0.0-alpha.1.19562.8 at [F:\workspace.9\_work\1\s\.dotnet\shared\Microsoft.NETCore.App]
[ERROR] 
[ERROR] You can resolve the problem by installing the specified framework and/or SDK.
[ERROR] 
[ERROR] The specified framework can be found at:
[ERROR]   - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=5.0.0-alpha1.19521.2&arch=x64&rid=win10-x64
```

JunTaoLuo commented:
I think we might need to add something like this to the interop tests to pick up the latest shared framework: https://github.com/aspnet/AspNetCore/blob/master/src/ProjectTemplates/test/Infrastructure/TemplateTests.props.in.

Essentially we need the workaround here: https://github.com/aspnet/AspNetCore/blob/master/Directory.Build.targets#L115-L122 to be able to run on the latest shared framework. However, I believe the repo has been "isolated" so that fix isn't picked up and the latest shared framework cannot be found.

JamesNK commented:
@JunTaoLuo Can you take this over - just need to get them passing in CI environment.

Commit:
Add gRPC interop tests to CI

- Convert to using references managed by build infrastructure
- Use produced AspNetCore.App shared framework
- Save server logs
- Dynamically bind to ports
- Ensure InteropWebsite is built in the same configuration as the test project

JunTaoLuo commented on src/Grpc/test/testassets/InteropClient/InteropClient.csproj:
```diff
@@ -0,0 +1,26 @@
+
+    <PackageReference Include="Grpc.Net.Client" Version="$(GrpcAspNetCorePackageVersion)" />
+
+    <PackageReference Include="CommandLineParser" Version="2.3.0" />
```
👀

JamesNK commented on src/Grpc/test/testassets/InteropClient/InteropClient.csproj:
```diff
@@ -1,11 +1,9 @@
   <PropertyGroup>
-    <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp5.0</TargetFramework>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
```
I made the interop client and website separate from the overall build infrastructure because they needed to reference the installed ASP.NET SDK, and not ASP.NET Core projects.

It appears you're integrating them into the overall build. How did you solve the ASP.NET SDK referencing issue?

JunTaoLuo replied:
By ASP.NET SDK do you mean the shared framework? So far I haven't done that yet, I'm only changing how the packages are referenced.

Though I see why we need to reference the shared framework in the InteropWebsite, why is it needed for the InteropClient? It seems to me that it should really just be a regular .NET Core app.

SteveSandersonMS commented:
The change to blazor.server.js looks unintentional - probably should be reverted.

JunTaoLuo commented:
Yup, that's accidental. Though I feel it's really easy to unintentionally modify that file.

JunTaoLuo commented:
It's now successfully building locally and on the CI. However the tests are only passing locally. On the CI it seems like the server process never ran. Also, I wonder why we didn't use the deployer pattern.

Commit:
Cleanup

JunTaoLuo commented:
The tests are up and running on the CI.

Tratcher commented:
This `dotnet run` approach doesn't work in Helix, right? We shouldn't be adding any new tests at this point that don't work in Helix.

JunTaoLuo commented:
It's unlikely that I'll be able to focus on this while I am working on extensions migration for the next couple of weeks. Given we have other tests in a similar position, we could leave the tests running on AzDO for now until we have a consistent story of how to test on Helix. I could potentially circle back to this when extensions work is complete.

HaoK commented:
dot net run approach will work on helix, as templates test generally require installing the shared framework bits onto the helix machine and doing dotnet new/run types of commands as written.  So gRPC tests should be able to do the same thing...that is once templates tests are fully working on helix

HaoK commented on src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs:
```diff
@@ -0,0 +1,85 @@
+        public WebsiteProcess(string path, ITestOutputHelper output)
+        {
+            var attributes = Assembly.GetExecutingAssembly()
+                .GetCustomAttributes<AssemblyMetadataAttribute>();
```
Oh, but we should revisit things like this, burning paths into the assembly doesn't work on helix, can we have this fallback to a relative path default that just works (that would be the helix path)

HaoK replied:
Like if the attribute doesn't exist, just fallback to "/artifacts/log" which is the default place we upload logs from on helix

JunTaoLuo replied:
The problem here isn't just about the path. Specifically, in this case, the test executable is trying to run the server/client executables but I don't think they are included in the Helix test payload. In general, we need a pattern for including other executables in test payloads.

However, in this particular case, I believe it might be possible to run the tests in process, in which case, we should be able to avoid the problem all together.

HaoK commented on src/Grpc/test/InteropTests/InteropTests.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            var attributes = Assembly.GetExecutingAssembly()
+                .GetCustomAttributes<AssemblyMetadataAttribute>();
+
```
Same thing, can we have a fallback (or better yet a config file),  @javiercn can you drive coming up with a pattern that we can use in place of this assembly metadata stuff?  Components and Templates are the main consumers of this metadata pattern I think right now

Commit:
Rebase fix

Commit:
Include tests assets in build directory for Helix

JunTaoLuo commented:
@HaoK can you take a quick look at the latest commit to see if it seems sane for Helix? I think I'll merge this once the Helix tests pass on CI.

HaoK commented on Directory.Build.targets:
```diff
@@ -167,5 +167,7 @@
   <Import Project="eng\targets\Npm.Common.targets"  Condition="'$(MSBuildProjectExtension)' == '.npmproj'" />
   <Import Project="eng\targets\ReferenceAssembly.targets" Condition=" $(HasReferenceAssembly) " />
   <Import Project="eng\targets\Helix.targets" Condition="'$(IsTestProject)' == 'true'" />
+  <Import Project="eng\targets\FunctionalTestAsset.targets" Condition="'$(IsTestAssetProject)' == 'true'" />
```
nit: You call it Integration test above, and functional test here.

HaoK commented on src/Shared/Process/ProcessEx.cs:
```diff
@@ -180,27 +174,22 @@ public void WaitForExit(bool assertSuccess, TimeSpan? timeSpan = null)
-        private static string GetNugetPackagesRestorePath() => (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("NUGET_RESTORE"))) 
-            ? typeof(ProcessEx).Assembly
+        private static string GetNugetPackagesRestorePath() =>
+            typeof(ProcessEx).Assembly
```
merge issues in this file, does this break template tests in helix?

JunTaoLuo replied:
Ah yes

HaoK commented on .azure/pipelines/ci.yml:
```diff
@@ -576,16 +576,18 @@ stages:
         displayName: Pack Templates
       - script: ./src/ProjectTemplates/build.cmd -ci -test -NoRestore -NoBuild -NoBuilddeps "/p:RunTemplateTests=true /bl:artifacts/log/template.test.binlog"
         displayName: Test Templates
+      - script: ./src/Grpc/build.cmd -ci -test "/bl:artifacts/log/grpc.test.binlog"
```
If grpc tests work on helix natively, I'd recommend not adding them to the azdo test legs, and leaving this as soley for Windows_templates_tests, and you can potentially remove any of the non helix codepaths (if any) as well then.

HaoK replied:
Actually I guess you still want the non helix code paths for local dotnet test runs

JunTaoLuo replied:
There's still a need to build the shared framework before we run these interop tests which is why we put the tests here. The goal here is to future proof them so it's easy to run them on Helix as @Tratcher suggested.

HaoK replied:
So these tests are already running on helix with this PR right?  I'm just asking why do we need to add these to the non helix test job if they are passing on helix.

JunTaoLuo replied:
Fair, I'll remove this once the tests are brought up on Helix.

Commit:
Incorporate changes in ProcessEx

Commit:
Include Grpc test in regular build

Commit:
Fixup

Commit:
Test

Commit:
exe doesn't always exist

Commit:
Merge branch 'master' into jamesnk/interoptests

Commit:
Capture logs on helix

Commit:
Maybe this will work

JunTaoLuo commented on src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs:
```diff
@@ -78,7 +78,17 @@ private void Process_OutputDataReceived(object sender, DataReceivedEventArgs e)
         public void Dispose()
         {
-            File.WriteAllText(_serverLogPath ?? Path.Combine(Directory.GetCurrentDirectory(), "InteropServer.log"), _consoleOut.ToString());
+            if (!string.IsNullOrEmpty(_serverLogPath))
```
I'll revert this commit if https://github.com/dotnet/aspnetcore/pull/20005#pullrequestreview-378116740 goes in.

JunTaoLuo replied:
@HaoK so eventually when #20005 goes in, it'll be correct to log directly to the current directory right?

HaoK replied:
No, you should write to artifacts/logs, might need to create that directory

JunTaoLuo replied:
so CurrentDirectory/artifacts/logs?

HaoK replied:
https://github.com/dotnet/aspnetcore/pull/20005 is finally merged, so maybe try one more rebase/build to see if the logs are working if you want

Commit:
There are two application started messages

Commit:
Derp

JunTaoLuo commented:
@HaoK The tests are passing on Helix! I'm doing one more pass to do a clean up to remove the AzDO modifications.

Commit:
Cleanup

HaoK commented:
Yup that should do it 

> On Mar 20, 2020, at 2:12 PM, John Luo <notifications@github.com> wrote:
> 
> ﻿
> @JunTaoLuo commented on this pull request.
> 
> In src/Grpc/test/InteropTests/Helpers/WebsiteProcess.cs:
> 
> > @@ -78,7 +78,17 @@ private void Process_OutputDataReceived(object sender, DataReceivedEventArgs e)
>  
>          public void Dispose()
>          {
> -            File.WriteAllText(_serverLogPath ?? Path.Combine(Directory.GetCurrentDirectory(), "InteropServer.log"), _consoleOut.ToString());
> +            if (!string.IsNullOrEmpty(_serverLogPath))
> so CurrentDirectory/artifacts/logs?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Commit:
Update directory

JunTaoLuo commented:
@JamesNK @dotnet/aspnet-build @HaoK I'm planning on merging this once tests pass. Anything else I should address before I do that?

JamesNK commented:
:shipit:

I would approve but it is my own PR 😄

HaoK commented on .azure/pipelines/ci.yml:
```diff
@@ -674,9 +674,9 @@ stages:
+      - script: .\restore.cmd -ci /p:BuildInteropProjects=true
         displayName: Restore
-      - script: .\build.cmd -ci -NoRestore -test -projects eng\helix\helix.proj /p:IsRequiredCheck=true /p:IsHelixJob=true /p:BuildAllProjects=true /p:BuildNative=true /p:RunTemplateTests=true /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log -bl
+      - script: .\build.cmd -ci -NoRestore -test -projects eng\helix\helix.proj /p:IsRequiredCheck=true /p:IsHelixJob=true /p:BuildAllProjects=true /p:BuildInteropProjects=true /p:BuildNative=true /p:RunTemplateTests=true /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log -bl
```
You will need to add this to all the various helix legs that you want these tests to run on (daily helix, and maybe the quarantined yml too since i imagine its possible some of these tests would get social distanced eventually too)

Commit:
Add interop tests to more pipelines

Commit:
mkdir
